pkg/controller/seed-controller-manager/encryption/controller.go

```diff
@@ -0,0 +1,290 @@
+		controllerutil.EnqueueClusterForNamespacedObject(mgr.GetClient()),
+		predicateutil.ByName(resources.EncryptionConfigurationSecretName),
+	); err != nil {
+		return fmt.Errorf("failed to create watcher for corev1.Secret: %w", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &appsv1.Deployment{}},
+		controllerutil.EnqueueClusterForNamespacedObject(mgr.GetClient()),
+		predicateutil.ByName(resources.ApiserverDeploymentName),
+	); err != nil {
+		return fmt.Errorf("failed to create watcher for appsv1.Deployment: %w", err)
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}, predicateutil.Factory(func(o ctrlruntimeclient.Object) bool {
+		cluster := o.(*kubermaticv1.Cluster)
+		encryptionEnabled := cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled
+		encryptionActive := cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionEncryptionInitialized, corev1.ConditionTrue)
+		return cluster.Spec.Features[kubermaticv1.ClusterFeatureEncryptionAtRest] && (encryptionEnabled || encryptionActive)
+	}))
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debug("Could not find cluster")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	log = log.With("cluster", cluster.Name)
+
+	// replicate the predicate from above to make sure that reconcile loops triggered by apiserver and secret
+	// do not run unexpected reconciles.
+	if !(cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled) && !(cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionEncryptionInitialized, corev1.ConditionTrue)) {
+		return reconcile.Result{}, nil
+	}
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionEncryptionControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(ctx, log, cluster)
+		},
+	)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// reconcile until encryption is successfully initialized
+	if cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled && !cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionEncryptionInitialized, corev1.ConditionTrue) {
```

Comment:
This line looks super duper similar to line 145, just the opposite. Can we put this magic logical expression into a function?

Reply:
It's actually not the same. The first one at line 145 is:

```go
!(cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled) && !cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionEncryptionInitialized, corev1.ConditionTrue)
```

Line 145 boils down to "if config in spec is not enabled and encryption was never initialised, ignore the cluster", while 175 is "if config in spec is enabled but encryption was not yet initialised, initialise the cluster". They're checking for different things, semantically.

Because the L145 check negates the cluster.Spec checks, I don't think we can easily move this to a function. The best we could do is

```go
func isEncryptionEnabled(cluster *kubermaticv1.Cluster) bool {
    return cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled
}
```

But I don't think that's worth doing, since the check itself is really simple?

Reply:
Never mind my comment above, this syncs nicely with https://github.com/kubermatic/kubermatic/pull/9654#discussion_r855291737. Consider it done.